lte/gateway/python/magma/pipelined/vlan_utils.py

```diff
@@ -0,0 +1,91 @@
+"""
+Copyright 2022 The Magma Authors.
+
+This source code is licensed under the BSD-style license found in the
+LICENSE file in the root directory of this source tree.
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+
+This file models common logic of the mandatory services ingress,
+middle and egress.
+"""
+
+from magma.pipelined.openflow import flows
+from magma.pipelined.openflow.magma_match import MagmaMatch
+from magma.pipelined.openflow.registers import TUN_PORT_REG, Direction
+
+
+def get_vlan_egress_flow_msgs(
+    dp, table_no, eth_type, ip, out_port=None,
+    priority=0, direction=Direction.IN, dst_mac=None,
+):
+    """
+    Install egress flows
+    Args:
+        dp datapath
+        table_no table to install flow
+        out_port specify egress port, if None reg value is used
+        priority flow priority
+        direction packet direction.
+    """
+    msgs = []
+    if out_port:
+        output_reg = None
+    else:
+        output_reg = TUN_PORT_REG
+
+    # Pass non vlan packet as it is.
+    # TODO: add support to match IPv6 address
+    if ip:
+        match = MagmaMatch(
+            direction=direction,
+            eth_type=eth_type,
+            vlan_vid=(0x0000, 0x1000),
+            ipv4_dst=ip,
+        )
+    else:
+        match = MagmaMatch(
+            direction=direction,
+            eth_type=eth_type,
+            vlan_vid=(0x0000, 0x1000),
+        )
+    actions = []
+    if dst_mac:
+        actions.append(dp.ofproto_parser.NXActionRegLoad2(dst='eth_dst', value=dst_mac))
+
+    msgs.append(
+        flows.get_add_output_flow_msg(
+            dp, table_no, match, actions,
+            priority=priority, output_reg=output_reg, output_port=out_port,
+        ),
+    )
+
+    # remove vlan header for out_port.
+    if ip:
+        match = MagmaMatch(
+            direction=direction,
+            eth_type=eth_type,
+            vlan_vid=(0x1000, 0x1000),
+            ipv4_dst=ip,
+        )
+    else:
+        match = MagmaMatch(
+            direction=direction,
+            eth_type=eth_type,
+            vlan_vid=(0x1000, 0x1000),
+        )
+    actions = [dp.ofproto_parser.OFPActionPopVlan()]
+    if dst_mac:
+        actions.append(dp.ofproto_parser.NXActionRegLoad2(dst='eth_dst', value=dst_mac))
+
+    msgs.append(
+        flows.get_add_output_flow_msg(
+            dp, table_no, match, actions,
+            priority=priority, output_reg=output_reg, output_port=out_port,
+        ),
+    )
+    return msgs
```

Comment:
did you ran non-nat datapath tests from S1AP integ tests with this PR?

Reply:
@pshelar, we run all tests in the "lte integ tests" workflow. Last run on fork: https://github.com/nstng/magma/runs/6115971885?check_suite_focus=true

Reply:
I dont see `test_attach_detach_non_nat_dp_ul_tcp.py` in the test log, can you check it.

Reply:
`test_attach_detach_non_nat_dp_ul_tcp.py` is in the `NON_SANITY_TESTS` list. It looks like the `NON_SANITY_TESTS` are not executed with the LTE integ-tests workflow. I do not see them executed anywhere automatically. Do you happen to know if this is intentionally?
I will execute them manually - but this will take some time.

Reply:
Lets move this test to 'PRECOMMIT_TESTS' set.
cc: @ssanadhya

Reply:
@pshelar executed the tests locally - `test_attach_detach_non_nat_dp_ul_tcp.py` was successful.

In general the tests seem to be very flaky (on a fresh vm setup). I still do not have all tests green. Do you have an opinion if one of the following tests might be relevant for the change?:
```
test_s1_handover_ping_pong.py
test_activate_deactivate_multiple_dedicated.py
test_sctp_shutdown_while_mme_is_stopped.py
test_secondary_pdn_with_dedicated_bearer_multiple_services_restart.py
test_3495_timer_for_dedicated_bearer_with_mme_restart.py
test_no_attach_complete_with_mme_restart.py
```

Reply:
these are not related datapath, so we can ignore it.